Board with player movement

shaylasry opened the pull request:
Created new Board and Player scriptable objects.
Created Player generator and player controller.
Implemented 2 movement approaches:
1. Character controller alone (use arrows for movement by default)
2. Character controller with Unity Input system (use W, S, A, D, or left joystick for movement)

Both know to handle floor collisions so the player will stop when reaching board boundaries.
Please notice that I used the 1.0f as ray distance for Physics.Raycast because the player Y coordinate is 1 and the board is 0.

lasrysha committed:
Board with player movement

itamarbiton commented on Assets/Animations/PlayerInput.cs:
Why is this in `Animations` folder? Create a new folder under `Assets` name it `Input` and put it there

itamarbiton commented on Assets/Animations/PlayerInput.cs.meta:
Move to `Input` folder

itamarbiton commented on Assets/Animations/PlayerInput.inputactions:
Move to `Input` folder

itamarbiton commented on Assets/Animations/PlayerInput.inputactions.meta:
Move to `Input` folder

itamarbiton commented on Assets/Data/Player.asset:
the only thing in this Scriptable object is the player's prefab? Why do we need this? can't we just use the prefab itself?

shaylasry replied:
For now, we can use the prefab. Made this so that if we want to make other related data like a username for the level selector we can use it.

itamarbiton commented on Assets/ScriptableObjects/Player.cs:
Again, the only thing in this scriptable object is the player's prefab? Is this really needed?

itamarbiton commented on Assets/Scripts/GameScene/IPlayerController.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public Board GetGameBoard();
+
+    private void HandleMovement()
+    {
+    }
+
+    private void HandleRotation()
+    {
+    }
+
+    private void HandleCollision()
+    {
+    }
+
+    private void HandlePhysics()
+    {
+    }
```
why do we need the brackets here? it's just a function declaration inside an interface, remove the brackets...

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    [SerializeField] private Board board;
+    private PlayerInput _playerInput;
+    private CharacterController _characterController;
+    
+    
```
why the double spacing? 1 line spacing is enough

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    private Vector2 _movementInput;
+    private Vector3 _playerMovement;
+    private float _playerSpeed = 2.0f;
+
+    
```
again remove the double spacing

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    
+    private Vector2 _movementInput;
+    private Vector3 _playerMovement;
+    private float _playerSpeed = 2.0f;
```
this should be configurable from outside of the script, make it `[SerializeField] private...`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        InitPlayerInput();
+        _characterController = GetComponent<CharacterController>();
+    }
+    
+
+ 
+
```
too many line breaks, remove it, one is enough

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    private void Awake()
+    {
+        InitPlayerInput();
+        _characterController = GetComponent<CharacterController>();
```
Let's try to refrain from using `GetComponent` as it creates a dependency between the script and the GameObject it's attached to, make `_characterController` a `[SerializeField] private...`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    private void Update()
+    {
+        HandleMovementWithPlayerInput();
+        // HandleMovement();
+        // _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
```
why are these comments here? remove them

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        _playerInput.PlayerContorls.Disable();
+    }
+
+    public Board GetGameBoard()
+    {
+        return AssetDatabase.LoadAssetAtPath<Board>("Assets/Data/Board.asset");
+    }
```
Let's try to refrain from doing this since it creates a dependency between the script and specific file in the project, remove this function and get the board's scriptable object as a `[SerializeField] private...`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        
+    }
+
+    private void HandleRotation()
+    {
+        throw new System.NotImplementedException();
+    }
+    
+    private void HandleCollision()
+    {
+        throw new System.NotImplementedException();
+    }
+
+    private void HandlePhysics()
+    {
+        throw new System.NotImplementedException();
+    }
```
why are these here? if they're not implemented remove them, no need for anything that's not implemented

itamarbiton replied:
what is this function supposed to do?

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    
+    private void HandleMovementWithPlayerInput()
+    {
+        // Vector3 moveDirection = new Vector3(_playerMovement);
```
remove the comment

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        Vector3 newPosition = transform.position + _playerMovement * Time.deltaTime * _playerSpeed;
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
```
why are we using Raycast here? Either use the board size, or explain why this is better

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    
+    private void HandleMovement()
+    {
+        
+        Vector3 moveDirection = new Vector3(Input.GetAxis("Horizontal"), 0f, Input.GetAxis("Vertical"));
+        moveDirection = transform.TransformDirection(moveDirection);
+        Vector3 newPosition = transform.position + moveDirection * Time.deltaTime * _playerSpeed;
+    
+      
+        RaycastHit hit;
+        float raycastDistance = 1f; 
+        if (Physics.Raycast(newPosition, Vector3.down, out hit, raycastDistance))
+        {
+            _characterController.Move(moveDirection * Time.deltaTime * _playerSpeed);
+        }
+        
+    }
```
what is the difference between this and `HandleMovementWithPlayerInput`? it's the same code duplicated! Either make it a function and call the function from both these methods or remove the one that we don't need

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+            _characterController.Move(_playerMovement * Time.deltaTime * _playerSpeed);
+        }
+    }
+    private void InitPlayerInput()
+    {
+        _playerInput = new PlayerInput();
+        
+        _playerInput.PlayerContorls.Movement.started += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.performed += OnMovementInput;
+        _playerInput.PlayerContorls.Movement.canceled += OnMovementInput;
+    }
```
You're registering the receive events but your not removing the registration when the object is disabled/destroyed, that's bad practice, please register inside `OnEnable` and unregister in `OnDisable`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+    {
+        _movementInput = context.ReadValue<Vector2>();
+
+        _playerMovement.x = _movementInput.x;
+        _playerMovement.y = 0.0f;
+        _playerMovement.z = _movementInput.y;
```
this can be done in 1 line of code, no need for 3... `_playerMovement = new Vector3(_movementInput.x, 0f, _movementInput.y);`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        _playerMovement.x = _movementInput.x;
+        _playerMovement.y = 0.0f;
+        _playerMovement.z = _movementInput.y;
+        _isMovementPressed = _movementInput.x != 0 || _movementInput.y != 0;
```
this can be done in a much more simple way... `_movementInput != Vector2.Zero`

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        _playerMovement.y = 0.0f;
+        _playerMovement.z = _movementInput.y;
+        _isMovementPressed = _movementInput.x != 0 || _movementInput.y != 0;
+        
```
Remove this empty line

itamarbiton commented on Assets/Scripts/GameScene/PlayerController.cs:
```diff
@@ -0,0 +1,136 @@
+        Vector3 playerDirection = new Vector3(_movementInput.x, 0.0f, _movementInput.y);
+        
+
+        Quaternion currentRotation = transform.rotation;
+        if (_isMovementPressed)
+        {
+            Quaternion targetRotation = Quaternion.LookRotation(playerDirection);
+            Quaternion.Slerp(currentRotation, targetRotation, rotationFactorPerFrame * Time.deltaTime);
+        }
+    }
```
If you can explain what's going on here I'll let you do it, but if not, please first understand what Quaternion.Slerp means before using it in this way

shaylasry replied:
Transform is a Quaternion object.

we use the Slerp function to rotate from the current gameObject.transform to the new direction we go to according to the input we got. The time it takes to finish the rotation is: rotationFactorPerFrame * Time.deltaTime

I can figure out another way if you think this one is just complicated and we can do it better.

itamarbiton commented on Assets/Scripts/GameScene/PlayerGenerator.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        Instantiate(player.prefab, new Vector3(0, 1, 0), Quaternion.identity);
+    }
+
+    
+    
```
Remove these empty lines

itamarbiton commented on Assets/Scripts/GameScene/PlayerGenerator.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    [SerializeField] private Player player;
+    
+
```
Remove this empty line

itamarbiton commented on Assets/Scripts/GameScene/PlayerGenerator.cs:
```diff
@@ -0,0 +1,22 @@
+
+using UnityEngine;
+
+public class PlayerGenerator : MonoBehaviour
```
Why do we need this script? What object uses it?

shaylasry replied:
If you mean the whole script file, Player Generator in Game Scene uses it the same as Board Generator uses BoardGenerator.cs

Do you want it to be in one merged file that generates both Board and Player?

itamarbiton commented on Logs/AssetImportWorker0-prev.log:
don't push log files in the repo, add them to the .gitignore

lasrysha committed:
Fixed Board movement

shaylasry committed:
Merge branch 'master' into Board-with-player-movment

shaylasry commented:
1. resolved all the fixed issues, for non-resolved please see my comments.
2. almost forgot, according to Rider IDE a private serializeFIled and public filed should start with lower case and private only should start with '_'. 
I followed this pattern but I know you don't like the use of '_'. Should we check this once and for all to keep on the same pattern as known in the industry?

itamarbiton committed:
Performed changes to the way the player moved

- Renamed Board to BoardData
- Renamed Player to PlayerData
- Handled movement in a more streamlined way

itamarbiton committed:
Removed unneeded code

lasrysha committed:
added new movement with board bounds

lasrysha committed:
added grass to board

lasrysha committed:
added coliision

lasrysha committed:
added mower and grass colission sound

lasrysha committed:
added grass tile

board still need some fixes

lasrysha committed:
added smoke

there are still board issues

IBLabs committed:
Merge pull request #3 from shaylasry/board-with-player-movement-improvements

Performed changes to the way the player moved
